### ISSUE TITLE
[Perf][B300] Fuse normalized linear-attention denominator epilogues

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/flash-linear-attention#15]
## Problem

The registered public `chunk_linear_attn` operator defaults to `normalize=True`. Its normalized path first computes the required key-prefix state and then evaluates the denominator and divide through an eager PyTorch chain in `StatefulNormalizeFunction`:

```python
denom = (q * scale * k_cum).sum(-1, keepdim=True) + 1e-10
o_out = o / denom
```

For bf16/fp16 this launches separate scale, multiply, reduction/add, and divide work and materializes K-sized intermediates after the prefix scan. The hand-written backward similarly evaluates the output divide, `do * o` reduction, denominator algebra, and two K-gradient chains as separate eager operations before the required reverse cumsum.

`chunk_linear_attn` and `fused_chunk_linear_attn` are explicitly decorated with `@torch.compiler.disable`, so the installed runtime is expected to expose this launch/traffic chain rather than capture it into one graph. `fused_recurrent_linear_attn` shares the helper but is not compiler-disabled; it is a secondary compatibility/control path, not part of the primary performance claim unless profiling independently confirms the same opportunity.

This is a B300 hypothesis, not a speedup claim. Profiling must first prove that the removable normalization fragment is material to the full registered endpoint.

## Proposed optimization

Add two private exact-SM103 Triton finalizers inside the linear-attention helper:

1. A forward program per token/head reduces the K dimension for the denominator and divides the V-vector, writing the normalized output and saved `[B,T,H,1]` denominator directly.
2. A backward program per token/head reduces `do * o` over V and writes `d_o_in`, `dq`, and `dk_cum` directly. The existing reverse `chunk_global_cumsum` remains responsible for final `dk`.

Keep both prefix scans, z-state extraction/gradient handling, numerator kernels, and public APIs unchanged. Do not fuse or edit `fla/ops/utils/cumsum.py`.

Numerical equivalence is a hard part of the design. The eager bf16/fp16 path rounds at operation boundaries. The finalizers must reproduce those boundaries explicitly rather than silently retaining all intermediates in FP32:

- forward: round after `q * scale`, after multiplication by `k_cum`, after the reduction result is cast to the input dtype, and after adding `1e-10` before dividing `o`;
- backward: preserve the dtypes/rounding of `do / denom`, `do * o`, the V reduction, `denom * denom`, the denominator division, `ddenom * scale`, and the final multiplication by `k_cum` or `q`.

Use `tl.sum`, not a dot path that changes TF32/accumulation policy. Every store must cover its full masked output, and all program-derived bases, strides, and offsets must use `tl.int64`.

## Bounded dispatch and fallback

The initial optimized dispatch is deliberately restricted to:

- exact NVIDIA capability `(10, 3)` via `fla.utils.get_device_capability`;
- bf16/fp16;
- `z_init is None`;
- `reverse is False`;
- dense or flattened packed-varlen training;
- profile-proven K/V bounds, expected to be at most 256.

The current implementation remains the exact fallback for non-SM103, fp32, state resume, reverse scans, unsupported dimensions/layouts, and measured losing buckets. Existing `dz_state` boundary additions, `dz_init`, z-state scatter/gather, and reverse-cumsum behavior stay outside the finalizers. Do not add a public backend, environment switch, architecture helper, shape lookup table, comparable replacement workspace, or changes to precision policy.

Expected tracked paths:

- `fla/ops/linear_attn/utils.py` for the finalizers and host dispatch;
- `tests/ops/test_linear_attn.py` only in an initial oracle-fixture commit, then frozen.

A new private file under `fla/ops/linear_attn/` is acceptable only if needed to keep `utils.py` readable; it may not widen the source scope. Do not edit benchmark sources, cumsum/common/simple-GLA kernels, public entry points, layers/models, or naive/reference implementations.

Lane-a issue #14 owns only preconditioned-KDA source and `tests/ops/test_precond_kda.py`; prior lane-b issue #13 is terminal RWKV6 work. Their tracked files and benchmark/model surfaces are disjoint from this issue.

## Honest public-path scope

The public operator and normalized chunk/fused-chunk layer modes are real user paths, but the shipped `LinearAttentionConfig` defaults `norm_feature_map=False`, and Based/ReBased default to parallel mode. Therefore this issue must not claim a default-model speedup.

Required integration evidence must instantiate a public normalized configuration explicitly, for example `LinearAttentionConfig(norm_feature_map=True, attn_mode="fused_chunk")`, or an equivalent public `LinearAttention`/Based/ReBased chunk-mode layer. Keep that reproducible harness outside git. The repository-native `chunk_linear_attn` benchmark remains the promotion endpoint because normalization is its public default.

## Fixture-first frozen correctness contract

Before production edits:

1. Run the complete untouched `tests/ops/test_linear_attn.py` on B300 at exact baseline `27967b970eaaf982a6960abf6cba8add9c34c7cc`. A deterministic red/OOR baseline is a terminal no-go.
2. Add focused positive-feature normalized cases against the existing naive recurrent oracle:
   - fp16 and bf16;
   - dense non-aligned T;
   - K classes 64/128/256;
   - at least one `K != V` case;
   - unequal packed varlen;
   - output plus every reachable `dq/dk/dv` gradient.
3. Validate this test-only change against untouched production and commit it as immutable `BENCH_BASE`.
4. Freeze the full test file, naive/reference code, tolerances, inputs, seeds, dtypes, skips, and numeric flags for every kernel iteration.

Existing fp32 normalized-gradient and split/resume tests remain mandatory fallback coverage. Add a non-contiguous or explicit stride case if the optimized dispatch accepts such layouts; otherwise prove the fallback handles it. Allocator NaN poisoning must remain enabled.

Per-iteration subsets are signal only. Promotion requires the full unfiltered gate:

```bash
python -m benchmarks.ops.verify --op chunk_linear_attn --base <BENCH_BASE> --modes fwd fwdbwd
python scripts/find_dependent_tests.py fla/ops/linear_attn/utils.py
```

Run every returned dependent test plus changed-file pre-commit/header/syntax checks and the banned-block-pointer sweep. No tolerance, reference, precision, skip, or environment-suppression edit is allowed after `BENCH_BASE`.

## Profile-first B300 gate

Before production edits, use torch.profiler or equivalent timeline evidence on untouched BF16 D64/D128/D256 dense rows and one packed-varlen row. Attribute the normalization fragment's individual launches, summed duration, transient allocation, and percentage of full public `chunk_linear_attn` fwd and fwdbwd.

Stop before implementation if the installed runtime has already fused the chain, if the removable fragment cannot plausibly support an 8% full-fwdbwd geomean, or if the path cannot preserve the eager dtype boundaries. The full prefix scans and numerator time are not removable and must be included in the Amdahl bound.

After the smallest proof finalizers, compile and profile D256 first. Stop rather than widen scope if reduction/vector live state causes material spills or occupancy collapse, or if launch savings are erased by the fused program.

## Repository-native benchmark plan

Use the existing `chunk_linear_attn` registry without editing benchmark files. Compare production HEAD with immutable `BENCH_BASE` in the same isolated B300 allocation:

```bash
python -m benchmarks.ops.verify --op chunk_linear_attn --base <BENCH_BASE> --modes fwd fwdbwd
python -m benchmarks.ops.run --op chunk_linear_attn --base <BENCH_BASE> --modes fwd fwdbwd
```

Run all six BF16 registry rows spanning D64/D128/D256. Add ignored, reproducible dense and unequal-packed public-op harnesses for `fused_chunk_linear_attn`; treat `fused_recurrent_linear_attn` as a correctness/non-regression control unless its actual execution boundary is profiled. `normalize=False`, fp32, state-resume, reverse, and unsupported-dimension calls are fallback controls.

For user integration, compare identical explicitly normalized public layer/model runs, dense and packed-varlen where the selected public endpoint supports them. Use the same config, seed, environment, warmup, steps, clocks, and allocation on baseline/candidate; do not report the default `linear_attn` model as exercising this optimization unless its config is explicitly changed to enable normalization.

Final evidence requires at least five interleaved same-session baseline/candidate rounds and each row's median, mean, standard deviation, min, p10, p20, p80, p90, sample count, and equal-weight geomean, plus peak allocation.

## NCU evidence

Collect both required Nsight Compute passes for candidate finalizers on representative D128 and D256 rows plus one packed-varlen endpoint:

- `--set full --section PmSampling --section PmSampling_WarpStates`
- `--set source --section SourceCounters`

Use timeline/kernel-sum evidence for the baseline eager fragment. Compare total fragment duration, launch count, DRAM bytes/throughput, L2 behavior, SM SOL, achieved occupancy, registers, local-memory spills, active/eligible warps, and dominant stalls. Confirm that both K-sized forward intermediates and the backward pointwise/reduction chain disappear without hidden casts, skipped outputs, changed prefix scans, or comparable replacement scratch.

Record exact baseline/candidate SHAs, commands, Slurm job IDs, GPU capability, clocks/idle evidence, and CUDA/PyTorch/Triton/NCU versions. Keep raw reports outside git under a cycle/lane-specific ignored profile directory.

## Acceptance and no-go conditions

- Full frozen fixture and dependent tests pass with unchanged API/layout/dtype and no numerical concessions.
- BF16 full `chunk_linear_attn` fwdbwd equal-weight geomean improves at least 1.08x across all six registered rows; full fwd geomean improves at least 1.05x.
- Every D64/D128/D256 class has a positive optimized-path row; no retained row regresses more than 3% outside demonstrated noise.
- Dense and packed normalized endpoints are non-regressing; the explicitly normalized public layer/model run improves at least 3% median throughput in one representative dense workload, with its packed counterpart neutral within 2% or better.
- Fallback and `normalize=False` controls remain within 2% noise, with no peak-memory regression.
- Profiler/NCU evidence attributes the gain to removed eager launches and transient traffic without spill, occupancy collapse, hidden precision changes, or skipped work.

Record no-go rather than widening scope if correctness requires relaxed tolerances, the runtime already fuses the chain, only a narrow artificial row wins, the full endpoint misses the bar, or the solution requires modifying shared cumsum/common kernels.

## Duplicate and mergeability audit

At upstream `27967b97`, no open upstream/fork issue, PR, or remote branch targets `StatefulNormalizeFunction`, `normalize_with_z_state`, or normalized linear-attention denominator epilogues. Merged #875 introduced the custom backward to restore the denominator-path dK gradient; it did not fuse these eager operations. Open #471 is a generic launch-optimization RFC with no LinearAttention implementation, branch, tracked files, or benchmark surface. Open #1144 changes benchmark input infrastructure for other operators and does not cover `chunk_linear_attn`; this issue edits no benchmark source.

If the gates pass, this is a focused upstream contribution: two private reduction/vector finalizers shared by normalized chunk paths, a registered dense/varlen public endpoint, explicit dtype-semantics preservation, exact SM103 fallback, and profiler-verifiable launch plus bandwidth savings without an API or model-format change.

**Comment by heiheiha798** [comment]
Cycle 14 is terminal as `PR_ABORTED` at exact draft-PR head `4fc3a94220f162361ac084af9a89a180d8401448`.

The earlier frozen fixture, direct helper, resource, and performance gates remain useful historical evidence, but fixed implementation review found that the shared SM103 normalization path was not validated through the reachable fused-recurrent caller. The approved review-resolution job 5418 exercised the actual fast path for a state-free positive-feature BF16 `fused_recurrent_linear_attn` case at B=1, T=35, H=2, K=64, V=80 against `naive_recurrent_linear_attn`. Backward `dq` failed the unchanged oracle at `2.124023% > 2.000000%`.

Durable evidence: `profile/c14-lane-b-linear-attn-norm-epilogues-opt/review-1/fla-c14b-review-val-5418.log` in the retained cycle-14 worktree.

The approved sequence stopped immediately on that correctness failure, so the FP16 fused-recurrent case, public normalized T<=64 route, dependent tests, repository-native frozen-base promotion commands, complete distribution report, and expanded D128/D256/packed NCU captures did not run. Fixed REVIEW returned `NOT_APPROVED` with all three findings still open and no new findings.

This exact head is not recommended for upstream submission. Issue #15 and draft PR #17 remain open as reproducible engineering records; the branch, commits, worktree, and evidence are preserved.
